### PR TITLE
[Profiler] Add queue depth computation

### PR DESCRIPTION
**Pull request by davidchencsl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80098
* #80094
* __->__ #79993
* #80159
* #79974

Test Plan:
Add test in test_profiler.py

### CODE REVIEW

File: test/test_profiler.py [hunk 1]
```diff
@@ -1147,10 +1147,21 @@ def test_utils_compute_queue_depth(self):
         # so kernel will queued up.
         # But later tensor indexing is overhead bound, and there
         # is sleep to make sure kernel finished before next dispatch.
-        golden_queue_depth_list = [1, 2, 3, 4, 5, 1, 1, 1]
+        golden_queue_depth_list = [
```
Comment (robieta) [inline test/test_profiler.py:1150]: I'm concerned by the magic numbers in the golden tests; it's very hard to actually verify if they make sense, or diagnose deviations. One thing that would make this easier to follow would be to stringify the expected result and then use expecttest. For example something like:
```
 0  [CPU]
 0  [cudaKernelLaunch]
 1  [CPU]
 1  [cudaKernelLaunch]
 2  [CPU]
 2  [cudaKernelLaunch]
    [GPU]
    [GPU]
    [GPU]
 0  [CPU]
...
```
(Feel free to play with the format.)

File: torch/profiler/_utils.py [hunk 2]
```diff
@@ -126,34 +127,44 @@ def is_cuda_kernel(e):
...
-                   cuda_launch_event.start_us() +
-                   cuda_launch_event.duration_us()):
+            if isinstance(event, _KinetoEvent):
+                start_time = event.start_us() * 1000
+                end_time = (event.start_us() + event.duration_us()) * 1000
+                # Find current spawned cuda kernel event
+                if event in kernel_mapping and kernel_mapping[
+                        event] is not None:
+                    spawned_kernel_index = kernel_mapping[event]
+            if isinstance(event, _ProfilerEvent):
```
Comment (robieta) [inline torch/profiler/_utils.py:152]: nit: `elif`

File: torch/profiler/_utils.py [hunk 3]
```diff
@@ -45,9 +46,9 @@ def intervals_overlap(self, intervals: List[Interval]):
         overlap_time = 0
         intervals = sorted(intervals, key=lambda x: x.start)
         for i, interval in enumerate(intervals):
-            if i + 1 < len(intervals):
-                assert interval.end <= intervals[
-                    i + 1].start, "Intervals must be disjoint"
+            if i + 1 < len(intervals) and interval.end > intervals[i +
```
Comment (robieta) [inline torch/profiler/_utils.py:49]: Why was this changed from an assert to a condition?

**Review by robieta — CHANGES_REQUESTED** [review]
Overall looks quite good. Very natural extension of the original algorithm, good use of pythonic constructs like `key=` to gracefully handle the mixed types. Once you address the test comment this should be good to go.

File: torch/profiler/_utils.py (hunk 3 above)
Comment (davidchencsl) [reply torch/profiler/_utils.py:49]: Because i actually came across two over lapping interval when I apply this in decrease interval. For the sake of convenience, I thought I could just change one of the interval's end to be non-overlapping.

Comment (davidchencsl) [reply torch/profiler/_utils.py]: Ah yes, that was what I was doing here. But I have reverted this change, because that only occurs after I introduce the decrease interval calculation. I will think about it in a later commit.

File: torch/profiler/_utils.py (hunk 2 above)
Comment (davidchencsl) [reply torch/profiler/_utils.py]: Done

File: test/test_profiler.py [hunk 4]
```diff
@@ -1234,6 +1234,45 @@ def test_utils_compute_self_time(self):
...
 
+    def test_utils_compute_queue_depth(self):
+
+        def format_queue_depth(queue_depth_list, events):
+            res = ""
+            for data, event in zip(queue_depth_list, events):
+                res += f"{data.queue_depth} [{event.name()}]\n"
+            return res
+
+        # We have to use Mock because time series data is too flakey to test
```
Comment (robieta) [inline test/test_profiler.py:1245]: nit: flaky

File: test/test_profiler.py [hunk 5]
```diff
@@ -1234,6 +1234,45 @@ def test_utils_compute_self_time(self):
...
+2 [GPU]
+1 [GPU]
+0 [GPU]
+""")
+        self.assertExpectedInline(
+            format_queue_depth([
+                basic_evaluation.metrics[k]
+                for k in basic_evaluation.event_keys
+            ], basic_evaluation.events), """\
+0 [CPU]
```
Comment (robieta) [inline test/test_profiler.py:1263]: I think it would be helpful to either give these more descriptive names or interleave the cuda events. For the first check it's pretty clear what's going on, but for this one it's less clear how to check for correctness.

File: torch/autograd/__init__.py [hunk 6]
```diff
@@ -333,7 +333,7 @@ def variable(*args, **kwargs):
                                 _supported_activities, _add_metadata_json, SavedTensor,
                                 _push_saved_tensors_default_hooks, _pop_saved_tensors_default_hooks)
 
-from torch._C._autograd import (_ProfilerResult, _KinetoEvent, _kineto_step,
+from torch._C._autograd import (_ProfilerResult, _KinetoEvent, _ProfilerEvent, _kineto_step,
```
Comment (robieta) [inline torch/autograd/__init__.py:336]: nit: as much as possible we want to limit the ties between profiler and the autograd namespace. (As it is only associated for legacy reasons.) Better to put this in `torch.profiler`

**Review by robieta — COMMENTED** [review]
Looks good, modulo the comment on the unit test. You've done a really good job managing the complexity that arose from mixing old and new style events.

File: torch/autograd/__init__.py (hunk 6 above)
Comment (davidchencsl) [reply torch/autograd/__init__.py:336]: Yes, I'll get rid of that, I no longer needed that symbol for type checking I think.

File: test/test_profiler.py (hunk 5 above)
Comment (davidchencsl) [reply test/test_profiler.py]: I annotate the cpu event name if that helps.

Comment (robieta) [reply test/test_profiler.py]: I think it's more that there's no way to check the correctness of the interleaving at a glance. (Because it depends on the specific timings.)

Comment (robieta) [reply test/test_profiler.py]: Actually, scratch that. The more descriptive names make it super clear exactly what's going on.

**Review by robieta — APPROVED** [review]
LGTM
